File: backend/routes/bookings.js
const express = require('express');
const { body, validationResult } = require('express-validator');
const Booking = require('../models/Booking');
const Destination = require('../models/Destination');
const auth = require('../middleware/auth');

const router = express.Router();

// @route   POST /api/bookings
// @desc    Create new booking
// @access  Private
router.post('/', auth, [
  body('destination_id').notEmpty(),
  body('check_in_date').isISO8601(),
  body('check_out_date').isISO8601(),
  body('guests').isInt({ min: 1, max: 20 }),
  body('special_requests').optional().isLength({ max: 500 })
], async (req, res) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ errors: errors.array() });
    }

    const { destination_id, check_in_date, check_out_date, guests, special_requests } = req.body;

    // Check if destination exists
    const destination = await Destination.findOne({ destination_id });
    if (!destination) {
      return res.status(404).json({ error: 'Destination not found' });
    }

    // Validate dates
    const checkIn = new Date(check_in_date);
    const checkOut = new Date(check_out_date);
    const today = new Date();
    
    if (checkIn < today) {
      return res.status(400).json({ error: 'Check-in date cannot be in the past' });
    }
    
    if (checkIn >= checkOut) {
      return res.status(400).json({ error: 'Check-out date must be after check-in date' });
    }

    // Calculate total price (simplified logic)
    const days = Math.ceil((checkOut - checkIn) / (1000 * 60 * 60 * 24));
    const basePrice = { '$': 100, '$$': 200, '$$$': 400, '$$$$': 800 }[destination.price_range] || 200;
    const total_price = basePrice * days * guests;

    const booking = new Booking({
      user_id: req.user.user_id,
      destination_id,
      check_in_date: checkIn,
      check_out_date: checkOut,
      guests,
      total_price,
      special_requests
    });

    await booking.save();
    
    res.status(201).json(booking);
  } catch (error) {
    console.error('Create booking error:', error);
    res.status(500).json({ error: 'Server error creating booking' });
  }
});

// @route   GET /api/bookings
// @desc    Get user bookings
// @access  Private
router.get('/', auth, async (req, res) => {
  try {
    const bookings = await Booking.find({ user_id: req.user.user_id })
      .sort({ created_at: -1 });
    
    // Add destination details to each booking
    const bookingsWithDestinations = await Promise.all(
      bookings.map(async (booking) => {
        const destination = await Destination.findOne({ 
          destination_id: booking.destination_id 
        });
        
        return {
          ...booking.toObject(),
          destination: destination ? {
            name: destination.name,
            city: destination.city,
            country: destination.country,
            images: destination.images.slice(0, 1)
          } : null
        };
      })
    );
    
    res.json(bookingsWithDestinations);
  } catch (error) {
    console.error('Get bookings error:', error);
    res.status(500).json({ error: 'Server error fetching bookings' });
  }
});

// @route   GET /api/bookings/:booking_id
// @desc    Get single booking
// @access  Private
router.get('/:booking_id', auth, async (req, res) => {
  try {
    const { booking_id } = req.params;
    
    const booking = await Booking.findOne({ 
      booking_id,
      user_id: req.user.user_id 
    });
    
    if (!booking) {
      return res.status(404).json({ error: 'Booking not found' });
    }
    
    // Add destination details
    const destination = await Destination.findOne({ 
      destination_id: booking.destination_id 
    });
    
    res.json({
      ...booking.toObject(),
      destination: destination ? {
        name: destination.name,
        city: destination.city,
        country: destination.country,
        images: destination.images,
        description: destination.description
      } : null
    });
  } catch (error) {
    console.error('Get booking error:', error);
    res.status(500).json({ error: 'Server error fetching booking' });
  }
});

// @route   PUT /api/bookings/:booking_id
// @desc    Update booking
// @access  Private
router.put('/:booking_id', auth, [
  body('check_in_date').optional().isISO8601(),
  body('check_out_date').optional().isISO8601(),
  body('guests').optional().isInt({ min: 1, max: 20 }),
  body('special_requests').optional().isLength({ max: 500 }),
  body('status').optional().isIn(['pending', 'confirmed', 'cancelled', 'completed'])
], async (req, res) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ errors: errors.array() });
    }

    const { booking_id } = req.params;
    const updates = req.body;

    const booking = await Booking.findOne({ 
      booking_id,
      user_id: req.user.user_id 
    });
    
    if (!booking) {
      return res.status(404).json({ error: 'Booking not found' });
    }

    // Update booking fields
    Object.keys(updates).forEach(key => {
      if (updates[key] !== undefined) {
        booking[key] = updates[key];
      }
    });

    await booking.save();
    
    res.json(booking);
  } catch (error) {
    console.error('Update booking error:', error);
    res.status(500).json({ error: 'Server error updating booking' });
  }
});

// @route   DELETE /api/bookings/:booking_id
// @desc    Cancel booking
// @access  Private
router.delete('/:booking_id', auth, async (req, res) => {
  try {
    const { booking_id } = req.params;
    
    const booking = await Booking.findOne({ 
      booking_id,
      user_id: req.user.user_id 
    });
    
    if (!booking) {
      return res.status(404).json({ error: 'Booking not found' });
    }

    booking.status = 'cancelled';
    await booking.save();
    
    res.json({ message: 'Booking cancelled successfully' });
  } catch (error) {
    console.error('Cancel booking error:', error);
    res.status(500).json({ error: 'Server error cancelling booking' });
  }
});

module.exports = router;

File: backend/routers/bookings.py
from fastapi import APIRouter, HTTPException, Depends
from typing import List
import uuid
from datetime import datetime

from models import BookingCreate, BookingResponse, BookingStatus
from database import get_database
from auth import get_current_user

router = APIRouter(prefix="/api/bookings", tags=["bookings"])

@router.post("/", response_model=BookingResponse)
async def create_booking(
    booking: BookingCreate,
    current_user: dict = Depends(get_current_user)
):
    """Create a new booking"""
    db = get_database()
    
    # Validate destination exists
    destination = await db.destinations.find_one({"destination_id": booking.destination_id})
    if not destination:
        raise HTTPException(status_code=404, detail="Destination not found")
    
    # Validate dates
    if booking.check_out_date <= booking.check_in_date:
        raise HTTPException(
            status_code=400, 
            detail="Check-out date must be after check-in date"
        )
    
    booking_data = {
        "booking_id": str(uuid.uuid4()),
        "user_id": current_user["user_id"],
        "destination_id": booking.destination_id,
        "destination_name": destination["name"],
        "check_in_date": booking.check_in_date,
        "check_out_date": booking.check_out_date,
        "guests": booking.guests,
        "special_requests": booking.special_requests,
        "status": BookingStatus.PENDING,
        "total_amount": 0.0,  # Calculate based on destination pricing
        "created_at": datetime.utcnow()
    }
    
    result = await db.bookings.insert_one(booking_data)
    booking_data["_id"] = str(result.inserted_id)
    
    return BookingResponse(**booking_data)

@router.get("/", response_model=List[BookingResponse])
async def get_user_bookings(current_user: dict = Depends(get_current_user)):
    """Get current user's bookings"""
    db = get_database()
    
    bookings = await db.bookings.find(
        {"user_id": current_user["user_id"]}
    ).to_list(length=100)
    
    result = []
    for booking in bookings:
        booking["_id"] = str(booking["_id"])
        
        # Add destination details
        destination = await db.destinations.find_one(
            {"destination_id": booking["destination_id"]}
        )
        if destination:
            booking["destination_details"] = {
                "name": destination["name"],
                "city": destination["city"],
                "country": destination["country"],
                "images": destination.get("images", [])
            }
        
        result.append(BookingResponse(**booking))
    
    return result

@router.delete("/{booking_id}")
async def cancel_booking(
    booking_id: str,
    current_user: dict = Depends(get_current_user)
):
    """Cancel a booking"""
    db = get_database()
    
    booking = await db.bookings.find_one({
        "booking_id": booking_id,
        "user_id": current_user["user_id"]
    })
    
    if not booking:
        raise HTTPException(status_code=404, detail="Booking not found")
    
    if booking["status"] != BookingStatus.PENDING:
        raise HTTPException(
            status_code=400, 
            detail="Only pending bookings can be cancelled"
        )
    
    await db.bookings.update_one(
        {"booking_id": booking_id},
        {"$set": {"status": BookingStatus.CANCELLED}}
    )
    
    return {"message": "Booking cancelled successfully"}